restructuredtext_lint/sphinx.py

```diff
@@ -0,0 +1,154 @@
+    from sphinx.domains import python as python_domain
+    from sphinx.domains import std as std_domain
+
+    SPHINX_AVAILABLE = True
+except ImportError:
+    SPHINX_AVAILABLE = False
+
+from docutils.parsers import rst
+from docutils.parsers.rst import directives as rst_directives
+from docutils.parsers.rst import roles as rst_roles
+
+# Default and/or base roles/directives to ignore.
+_base_sp_roles = tuple(['ctype'])
+_base_sp_directives = tuple(['autosummary',
+                             'centered', 'currentmodule', 'deprecated',
+                             'hlist', 'include', 'index',
+                             'literalinclude', 'no-code-block',
+                             'seealso', 'toctree', 'todo',
+                             'versionadded', 'versionchanged'])
+
+
+class _IgnoredDirective(rst.Directive):
+    """Stub for unknown directives.
+
+    See: https://github.com/myint/rstcheck/blob/master/rstcheck.py#L305
+    """
+    has_content = True
+
+    def run(self):
+        return []
+
+
+def _ignore_role(name, rawtext, text, lineno, inliner,
+                 options=None, content=None):
+    """Stub for unknown roles.
+
+    See: https://github.com/myint/rstcheck/blob/master/rstcheck.py#L316
+    """
+    return ([], [])
+
+
+def _fetch_roles_directives(no_skip_directives=None, no_skip_roles=None):
+    # Extract all possible directives & roles that sphinx is aware of for
+    # all known sphinx domains, and then remove the ones we do not want to
+    # skip (and return them).
+    sp_directives = list(_base_sp_directives)
+    sp_directives.extend(std_domain.StandardDomain.directives)
+    sp_roles = list(_base_sp_roles)
+    sp_roles.extend(std_domain.StandardDomain.roles)
+
+    for (domain, class_name) in [(c_domain, 'CDomain'),
+                                 (cpp_domain, 'CPPDomain'),
+                                 (javascript_domain, 'JavaScriptDomain'),
+                                 (python_domain, 'PythonDomain')]:
+
+        # Get all the domains directives and roles and insert them.
+        domain_class = getattr(domain, class_name)
+        domain_directives = getattr(domain_class, 'directives', [])
+        domain_roles = getattr(domain_class, 'roles', [])
+
+        sp_directives.extend(domain_directives)
+
+        # Ensure that we also use the name prefixed version as well
+        # for example :py:func: and :func: are equivalent and we need to make
+        # sure we register both kinds.
+        sp_directives.extend('%s:%s' % (domain_class.name, item)
+                             for item in domain_directives)
+
+        sp_roles.extend(domain_roles)
+        sp_roles.extend('%s:%s' % (domain_class.name, item)
+                        for item in domain_roles)
+
+    # Filter out the ones we desire to avoid skipping (if any)...
+    if not no_skip_directives:
+        no_skip_directives = []
+    if not no_skip_roles:
+        no_skip_roles = []
```

Comment:
Actually, the user is never passing in anything about skipping from the main file. We should ditch this entirely. If they want to build their own directives with skips, then they should be allowed to. However, we should not complicate our library for their usage. If anything, that can be a third party module to build out said directives.